### ISSUE TITLE
CMake test framework plugin

### ISSUE THREAD

**Issue by LecrisUT** [teemtee/tmt#2758]
I want to sketch out some ideas for making tmt-cmake plugin:
- [ ] overall: how to write a tmt plugin? E.g. how can it be included in the `.fmf` files, how to overwrite it via an editable-like installation, how to minimize coupling with upstream
- [ ] overall: plugin will make sure CMake is available (maybe also ninja for good measure) from package manager
- [ ] `prepare` plugin: runs a configure-build (maybe also install) preset/traditional workflow
- [ ] `test.framework` plugin: run `ctest` for a build environment defined in previous feature
- [ ] `test.discover` plugin: convert test names and `ctest` labels to tmt tests and tags
- [ ] `test.report` plugin: recover junit results from ctest and forward it down
- [ ] CMake: make tmt commands and/or python environment available in CMake
- [ ] CMake: add test wrapper to convert tmt-native test to ctest callable tests
- [ ] CMake: expose a tmt subcommand to be executed as part of `ctest --build-and-test`. E.g. thic could parse the build stdout/stderr, build folder artifacts, etc. as part of a regression test and pass it to something like beakerlib.

**Comment by lukaszachy** [comment]
Hopefully  https://tmt.readthedocs.io/en/stable/code/plugin-introduction.html answers the 'how to write plugin' questions. I don't see the framework documented, so this is https://github.com/teemtee/tmt/blob/main/tmt/frameworks/__init__.py

tldr; it isn't in `.fmf` files but you need to have installed on the runner.

**Comment by happz** [comment]
Indeed, WRT test framework https://github.com/teemtee/tmt/blob/main/tmt/frameworks/__init__.py#L38 (https://tmt.readthedocs.io/en/stable/code/autodocs/tmt.frameworks.html#tmt.frameworks.TestFramework) and e.g. https://github.com/teemtee/tmt/blob/main/tmt/frameworks/shell.py

WRT prepare plugin, any would work as an example, e.g. https://github.com/teemtee/tmt/blob/main/tmt/steps/prepare/shell.py

**Comment by LecrisUT** [comment]
Thanks, what I had in mind is to have a `requirement.txt` section for the `fmf` files where the plugins can auto-magically be pulled. Probably always on the `plans` level. Since they all have to be python based, we can just call `pip` internally to get the dependencies, with probably a safe-guard that the automation only occurs in `venv`, and outside of it there is a print-out of where/how to get the dependency from pypi.

One complication in that approach is how to refresh the python environment to load the newly installed packages :thinking:

**Comment by lukaszachy** [comment]
'auto-magicall' pull would be interesting to have, but I'm afraid that for production CI it will be disabled (or allow-listed) anyway so I'm not sure that it is worth pursuing right now. Lets' start with cmake as a plugin which is pip installed manually (read: on workers as TF or actions or ..)

We have 2 plugins internally and frankly speaking it is pain to keep them working with 3 versions of tmt - the released (pypi, fedora), the main branch (copr teemtee/tmt for early adopters) and PR which modifies (read: breaks) API.
Code of those plugins can have lot of `hasattr` if/else if some release is particular heavy with development.

With that said, having it this repo is easier for the maintainer of such plugin.

**Comment by LecrisUT** [comment]
I have started the repo: https://github.com/LecrisUT/tmt-cmake and I'll see how it goes. First step is to get the plugin to be loaded and detected when `tmt plans show`

**Comment by LecrisUT** [comment]
Seeing the light at the end of the tunnel, the only remaining part to have a minimum plugin is extracting the test results more cleanly and adding more tests. Will soon ping for a bit of review after that.

**Comment by psss** [comment]
@LecrisUT, let's give a short intro about the goal here on the triage meeting?

**Comment by LecrisUT** [comment]
Could probably just close this. The proof of concept is there, just need to chip away at it. It would be useful to work on this one a bit more because it does uncover feature request needs for plugin creators, and how to deal with `discover`-`prepare`-`framework` combos in general.
